File: src/net/opgenorth/yeg/buildings/widget/BuildingRowWrapper.java
package net.opgenorth.yeg.buildings.widget;

import android.view.View;
import android.widget.TextView;
import net.opgenorth.yeg.buildings.R;
import net.opgenorth.yeg.buildings.model.RelativeBuildingLocation;

import java.text.DecimalFormat;

public class BuildingRowWrapper {
	private View _base = null;
	private TextView _nameLabel;
	private TextView _addressLabel;
	private TextView _yearBuiltLabel;
	private TextView _distanceToMeLabel;

	public BuildingRowWrapper(View base) {
		_base = base;
	}

	public TextView getNameLabel() {
		if (_nameLabel == null) {
			_nameLabel = (TextView) _base.findViewById(R.id.building_row_name);
		}
		return _nameLabel;
	}

	public TextView getAddressLabel() {
		if (_addressLabel == null) {
			_addressLabel = (TextView) _base.findViewById(R.id.building_row_address);
		}
		return _addressLabel;
	}

	public TextView getYearBuiltLabel() {
		if (_yearBuiltLabel == null) {
			_yearBuiltLabel = (TextView) _base.findViewById(R.id.building_row_year_built);
		}
		return _yearBuiltLabel;
	}

	public TextView getDistanceToMeLabel() {
		if (_distanceToMeLabel == null) {
			_distanceToMeLabel = (TextView) _base.findViewById(R.id.building_distance_to_me);
		}
		return _distanceToMeLabel;
	}

	public void display(RelativeBuildingLocation relativeBuilding) {
		getNameLabel().setText(relativeBuilding.getBuilding().getName());
		getAddressLabel().setText(relativeBuilding.getBuilding().getAddress());
		getYearBuiltLabel().setText("Construction date: " + relativeBuilding.getBuilding().getConstructionDate());

		if (relativeBuilding.getDistance() < 1) {
			getDistanceToMeLabel().setVisibility(View.GONE);
		}
		else {
			// TODO  refactor out the number formatting.
			DecimalFormat formatter;
			String units;
			double distance = relativeBuilding.getDistance();
			if (distance < 1000) {
				units = " m.";
				formatter = new DecimalFormat("####");
			}
			else {
				units = " km.";
				distance = distance / 1000;
				if (distance < 100) {
					formatter = new DecimalFormat("####.#");
				}
				else {
					formatter = new DecimalFormat("#####");
				}
			}

			getDistanceToMeLabel().setVisibility(View.VISIBLE);
			getDistanceToMeLabel().setText("Distance: " + formatter.format(distance) + units);
		}
	}
}

File: dotnet/src/yegbuildings/Widgets/BuildingRowWrapper.cs
using Android.Views;
using Android.Widget;
using Java.Lang;
using Net.Opgenorth.Yeg.Buildings.Model;

namespace Net.Opgenorth.Yeg.Buildings.Widgets
{
    public class BuildingRowWrapper : Object
    {
        private readonly View _baseView;
        private TextView _nameLabel;
        private TextView _addressLabel;
        private TextView _yearBuiltLabel;
        private TextView _distanceToMeLabel;

        public BuildingRowWrapper(View baseView)
        {
            _baseView = baseView;
            _baseView.Tag = this;
        }

        private TextView NameLabel
        {
            get { return _nameLabel ?? (_nameLabel = _baseView.FindViewById<TextView>(Resource.Id.building_row_name)); }
        }

        private TextView AddressLabel
        {
            get { return _addressLabel ?? (_addressLabel = _baseView.FindViewById<TextView>(Resource.Id.building_row_address)); }
        }

        private TextView YearBuilt
        {
            get { return _yearBuiltLabel ?? (_yearBuiltLabel = _baseView.FindViewById<TextView>(Resource.Id.building_row_year_built)); }
        }

        private TextView DistanceToMeLabel
        {
            get { return _distanceToMeLabel ?? (_distanceToMeLabel = _baseView.FindViewById<TextView>(Resource.Id.building_distance_to_me)); }
        }

        public void Display(RelativeBuildingLocation relativeBuilding)
        {
            NameLabel.Text = ((Building) relativeBuilding).Name;
            AddressLabel.Text = ((Building) relativeBuilding).Address;
            YearBuilt.Text = "Construction Date: " + ((Building) relativeBuilding).ConstructionDate;
            const string units = " km.";
            if (relativeBuilding.Distance < 1)
            {
                DistanceToMeLabel.Visibility = ViewStates.Gone;
                DistanceToMeLabel.Text = string.Empty;
            }
            else
            {
                DistanceToMeLabel.Visibility = ViewStates.Visible;
                var distance = relativeBuilding.Distance/1000.0;
                var formatString = (distance < 100) ? "{0:0000.0}" : "{0:00000}";
                DistanceToMeLabel.Text = "Distance: " + string.Format(formatString, distance) + units;
            }
        }
    }
}